ls1intum/Artemis
# `Development`: Refactor the relation between learning goals and exercises

### Checklist
#### General


#### Server

### Motivation and Context



Instructors can create _lecture units_ for a lecture, where they may also choose to add existing exercises as a lecture unit (so called "exercise units").
This is a special case, because title and release date for an _exercise unit_ are taken from the connected exercise.
Instructors may also connect lecture units, including exercise units, to one or more learning goals (this relation is saved in the “learning_goal_lecture_unit” table).

However, when a course does not use lecture units, there exists no way to connect _exercises_ to _learning goals_.

### Description


In this PR, we refactor the relation between _learning goals_ and _exercises_ to **eventually** allow instructors to connect _exercises_ to _learning goals_.
The database table and class attributes already existed (“learning_goal_exercise” table), but were previously not used.
The migration included in this PR moves existing relations between _exercise units_ and _learning goals_ to this table and removes them from the “learning_goal_lecture_unit” table.
**Now, the learning goals for an exercise unit are taken from the connected exercise.**
You should not notice any changes in the client/as an instructor. Client changes for learning goals are coming up in a future PR, so a user interface to connect exercises and learning goals is not provided, yet.

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Log in to Artemis
2. Navigate to Course Administration
3. Go to Learning Goals
4. Ensure you can still create and delete learning goals
5. Make sure you can edit learning goals and connect them to all kids of lecture units (especially exercise units)
6. Also test creating, editing, deleting lectures, exercises, and lecture units

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Instructions | Branches |
|------------|-------:|-----:|
| LearningGoalService.java | 95% | 62% |
| LectureService.java | 95% | 91% |
| LectureUnitService.java | 90% | 83% |
| LearningGoalResource.java | 92% | 63% |
| LectureResource.java | 97% | 78% |

### Screenshots



<img src="https://user-images.githubusercontent.com/102237363/189896930-d552a43f-97c3-499f-88a0-50e05298e378.png" width="600">
